Logging - fixes and optimizations

dmatej opened the pull request:
* Fixes #25349 - Fixed unstable LogFileManager logging
* Preferring onSpinWait over yield
* Added missing serialVersionUUID to CapacitySemaphore
* Overriden default implementations of SystemLogger (resolved message supplier too early)

dmatej committed:
Stabilized LogFileManager logging and class+method resolution

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Preferring onSpinWait over yield

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Added missing serialVersionUID to CapacitySemaphore

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Optimized GlassFishLoggerFinder - default methods resolved the supplier early

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej commented:
First run failed, second passed, but I cannot reproduce it and even explain it, at least now. However it should be still better than before.
```
2025-02-08T02:41:28.0616458Z [ERROR] Tests run: 4, Failures: 2, Errors: 0, Skipped: 0, Time elapsed: 0.108 s <<< FAILURE! -- in org.glassfish.main.jul.handler.GlassFishLogHandlerTest
2025-02-08T02:41:28.0617305Z [ERROR] org.glassfish.main.jul.handler.GlassFishLogHandlerTest.enableStandardStreamsLoggers(TestInfo) -- Time elapsed: 0.056 s <<< FAILURE!
2025-02-08T02:41:28.0617524Z org.opentest4j.MultipleFailuresError: 
2025-02-08T02:41:28.0617684Z Multiple Failures (1 failure)
2025-02-08T02:41:28.0617872Z 	java.lang.AssertionError: file content
2025-02-08T02:41:28.0620896Z Expected: iterable containing [a string containing "INFO", "main", "Tommy, can you hear me?" in order, a string containing "INFO", "main", "org.glassfish.main.jul.handler.GlassFishLogHandlerTest.enableStandardStreamsLoggers", "Some info message" in order, a string containing "SEVERE", "main", "Can you feel me near you?" in order, a string containing "SEVERE", "main", "Příliš žluťoučký kůň úpěl ďábelské ódy" in order]
2025-02-08T02:41:28.0621298Z      but: no item was a string containing "INFO", "main", "Tommy, can you hear me?" in order
2025-02-08T02:41:28.0621606Z 	at org.junit.jupiter.api.AssertAll.assertAll(AssertAll.java:80)
2025-02-08T02:41:28.0621913Z 	at org.junit.jupiter.api.AssertAll.assertAll(AssertAll.java:44)
2025-02-08T02:41:28.0622225Z 	at org.junit.jupiter.api.AssertAll.assertAll(AssertAll.java:38)
2025-02-08T02:41:28.0622560Z 	at org.junit.jupiter.api.Assertions.assertAll(Assertions.java:2944)
2025-02-08T02:41:28.0623475Z 	at org.glassfish.main.jul@7.0.23-SNAPSHOT/org.glassfish.main.jul.handler.GlassFishLogHandlerTest.enableStandardStreamsLoggers(GlassFishLogHandlerTest.java:115)
2025-02-08T02:41:28.0623732Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
2025-02-08T02:41:28.0623983Z 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
2025-02-08T02:41:28.0624389Z 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
2025-02-08T02:41:28.0624598Z 	Suppressed: java.lang.AssertionError: file content
2025-02-08T02:41:28.0627316Z Expected: iterable containing [a string containing "INFO", "main", "Tommy, can you hear me?" in order, a string containing "INFO", "main", "org.glassfish.main.jul.handler.GlassFishLogHandlerTest.enableStandardStreamsLoggers", "Some info message" in order, a string containing "SEVERE", "main", "Can you feel me near you?" in order, a string containing "SEVERE", "main", "Příliš žluťoučký kůň úpěl ďábelské ódy" in order]
2025-02-08T02:41:28.0627900Z      but: no item was a string containing "INFO", "main", "Tommy, can you hear me?" in order
2025-02-08T02:41:28.0628182Z 		at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
2025-02-08T02:41:28.0629091Z 		at org.glassfish.main.jul@7.0.23-SNAPSHOT/org.glassfish.main.jul.handler.GlassFishLogHandlerTest.lambda$enableStandardStreamsLoggers$7(GlassFishLogHandlerTest.java:118)
2025-02-08T02:41:28.0629416Z 		at org.junit.jupiter.api.AssertAll.lambda$assertAll$0(AssertAll.java:68)
2025-02-08T02:41:28.0629838Z 		at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
2025-02-08T02:41:28.0630393Z 		at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:1024)
2025-02-08T02:41:28.0630819Z 		at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
2025-02-08T02:41:28.0631501Z 		at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
2025-02-08T02:41:28.0631957Z 		at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
2025-02-08T02:41:28.0632392Z 		at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
2025-02-08T02:41:28.0632779Z 		at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
2025-02-08T02:41:28.0633057Z 		at org.junit.jupiter.api.AssertAll.assertAll(AssertAll.java:77)
2025-02-08T02:41:28.0633166Z 		... 7 more
2025-02-08T02:41:28.0633176Z 
2025-02-08T02:41:28.0633695Z [ERROR] org.glassfish.main.jul.handler.GlassFishLogHandlerTest.roll -- Time elapsed: 0.022 s <<< FAILURE!
2025-02-08T02:41:28.0633870Z org.opentest4j.MultipleFailuresError: 
2025-02-08T02:41:28.0634005Z Multiple Failures (1 failure)
2025-02-08T02:41:28.0634336Z 	java.lang.AssertionError: file content
2025-02-08T02:41:28.0636557Z Expected: iterable containing [a string containing "INFO", "main", "Tommy, can you hear me?" in order, a string containing "INFO", "main", "Some info message" in order, a string containing "SEVERE", "main", "Can you feel me near you?" in order, a string containing "SEVERE", "main", "Příliš žluťoučký kůň úpěl ďábelské ódy" in order, a string containing "SEVERE", "main", "File one, record one" in order]
2025-02-08T02:41:28.0636953Z      but: no item was a string containing "INFO", "main", "Tommy, can you hear me?" in order
2025-02-08T02:41:28.0637264Z 	at org.junit.jupiter.api.AssertAll.assertAll(AssertAll.java:80)
2025-02-08T02:41:28.0637569Z 	at org.junit.jupiter.api.AssertAll.assertAll(AssertAll.java:44)
2025-02-08T02:41:28.0637877Z 	at org.junit.jupiter.api.AssertAll.assertAll(AssertAll.java:38)
2025-02-08T02:41:28.0638208Z 	at org.junit.jupiter.api.Assertions.assertAll(Assertions.java:2944)
2025-02-08T02:41:28.0638990Z 	at org.glassfish.main.jul@7.0.23-SNAPSHOT/org.glassfish.main.jul.handler.GlassFishLogHandlerTest.roll(GlassFishLogHandlerTest.java:139)
2025-02-08T02:41:28.0639306Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
2025-02-08T02:41:28.0639577Z 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
2025-02-08T02:41:28.0639839Z 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
2025-02-08T02:41:28.0640057Z 	Suppressed: java.lang.AssertionError: file content
2025-02-08T02:41:28.0642419Z Expected: iterable containing [a string containing "INFO", "main", "Tommy, can you hear me?" in order, a string containing "INFO", "main", "Some info message" in order, a string containing "SEVERE", "main", "Can you feel me near you?" in order, a string containing "SEVERE", "main", "Příliš žluťoučký kůň úpěl ďábelské ódy" in order, a string containing "SEVERE", "main", "File one, record one" in order]
2025-02-08T02:41:28.0642810Z      but: no item was a string containing "INFO", "main", "Tommy, can you hear me?" in order
2025-02-08T02:41:28.0643322Z 		at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
2025-02-08T02:41:28.0644355Z 		at org.glassfish.main.jul@7.0.23-SNAPSHOT/org.glassfish.main.jul.handler.GlassFishLogHandlerTest.lambda$roll$10(GlassFishLogHandlerTest.java:142)
2025-02-08T02:41:28.0644711Z 		at org.junit.jupiter.api.AssertAll.lambda$assertAll$0(AssertAll.java:68)
2025-02-08T02:41:28.0645122Z 		at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
2025-02-08T02:41:28.0645601Z 		at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:1024)
2025-02-08T02:41:28.0646019Z 		at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
2025-02-08T02:41:28.0646485Z 		at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
2025-02-08T02:41:28.0646905Z 		at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
2025-02-08T02:41:28.0647315Z 		at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
2025-02-08T02:41:28.0647729Z 		at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
2025-02-08T02:41:28.0648216Z 		at org.junit.jupiter.api.AssertAll.assertAll(AssertAll.java:77)
2025-02-08T02:41:28.0648353Z 		... 7 more
```

OndroMih commented:
> First run failed, second passed

@dmatej, I observed something similar - when I run the tests for the `glassfish-jul-extension` project, with `mvn test`, I randomly get the same error. I was getting it also before, without your changes, or with my changes in https://github.com/eclipse-ee4j/glassfish/pull/25365, so I think it's not related. It must be some race condition, which causes that the log messages are never logged.

OndroMih commented:
I found out a little more, although not the root cause. When I deleted the line with "logFile.deleteOnExit();", so that the log file is kept, I see that the rolling didn't happen when the test failed. There's a single log file, which contains the "Tommy, can you hear me?" and other messages, and no log_DATE file

That most probably means that the file was written after the assertion. What I think is appening is that, messages are logged, then assertion checks if they are in the file, the messages are still being processed asynchronously and not yet in the file, and then the assertion fails. The messages are eventually written to the file later. So, after the tests fail, the messages are in the log file. They are not in the rolled file, because the roll test method first checks if the messages are present in the log. They are still not present, so also the roll test fails at the beginning, without  calling the roll method.

I saw this happening also when I debugged the test - I got to a point when the assertion failed. At that time, I checked the log and it was empty. When I reached the AfterAll method, which closes the log file, I checked the file again, and the messages were there.

So, I believe, that eventually messages get written to the log, but they are delayed. When I increased the pump sleep to 2 seconds, it helped a bit. My script ran much more iterations without a failure, but then still failed after some time. So it seems that sometimes the log messages are delayed even more than 2 seconds, which is probably some deadlock or race condition.

dmatej commented:
I can reproduce it on CI in some 50% of runs, but locally it did not fail even once until I set the sleep time just to 1 nano, haha. You idea to comment out the `deleteOnExit` call was what I needed to understand what is happening.

The problem is that while id of log record is atomic sequence, timestamp is set on creation, BUT that doesn't mean that one or the other concurrent thread will keep the order. The order of records in the file cannot be reliable, we would have to wait in handler and slowly sort records before writing - basically we could log just records waiting for some time in the queue.

It is not a bug, it is a feature, however I will try yet to 
* make the failure more probable
* face the fact that the ordering can change and it is ok

dmatej committed:
Stabilized GlassFishLogHandlerTest

- Concurrent creation and logging of log records cannot have predictable order
- So we are waiting until the file grows to the expected number of lines
  instead of sleep which might or might not be enough.

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>
